### ISSUE TITLE
Dockerfile missing?

### ISSUE THREAD

**Issue by bherbruck** [devcontainers/templates#77]

**Comment by andreujuanc** [comment]
Can we get an option to add a dockerfile that exists on local? So if I had done `docker pull node:latest` then when I run the VSCode command to add a devContainer to it there could be an option that shows "locally available images" and so I can just chose from there. :D

**Comment by swfsql** [comment]
There is a small workaround to make it easier for you to use your plain and own Dockerfile. 

Eg. your original Dockerfile:
```Dockerfile
FROM python:alpine3.17
# (Other commands that you might have..)
```

Instead of using it directly, you can add a header and a footer alpine configuration that is required by the remote vs-code.

`.devcontainer/devcontainer.json`:
```
{
	"name": "XYZ",
	"build": {
		"dockerfile": "Dockerfile",
		"context": "."
	}
}
```

`.devcontainer/Dockerfile`:
```Dockerfile
# vscode devcontainer configuration header
FROM alpine/git:latest AS devcontainers-images
WORKDIR /tmp/git/
RUN git clone https://github.com/devcontainers/images.git

# your actual dockerfile content
FROM python:alpine3.17
# (Other commands that you might have..)

# vscode devcontainer configuration footer
# Reference: https://github.com/devcontainers/images/blob/main/src/base-alpine/
COPY --from=devcontainers-images /tmp/git/images/src/base-alpine/.devcontainer/library-scripts/common-alpine.sh  /tmp/library-scripts/common-alpine.sh
ARG INSTALL_ZSH="true"
ARG USERNAME=vscode
ARG USER_UID=1000
ARG USER_GID=$USER_UID
RUN apk update &&\ 
    ash /tmp/library-scripts/common-alpine.sh "${INSTALL_ZSH}" "${USERNAME}" "${USER_UID}" "${USER_GID}" &&\
    rm -rf /tmp/library-scripts
```

In this way you end up with any base Dockerfile that you might have (alpine with python, directly from alpine in this case) and still have the needed vscode-remote configuration. I only checked the alpine one, but for other bases there could have different configuration/argument needs, so you should check them out accordingly.

**Comment by Chuxel** [comment]
@andreujuanc That would end up being specific to the Dev Containers extension I suspect since it would need access to your local machine and docker would have to be installed.  But we could do that in concept - or we could just have an option to enter an image name as a starting point.

I do think it would be pretty easy to automate the process of an `image` reference getting converted into a stub Dockerfile or Docker Compose file here. @burkeholland also mentioned being confused by them not being present, and it impacts docs and samples.

**Comment by andreujuanc** [comment]
> Just to clarify, Dockerfiles are still fully supported. It's just more an an advanced use case, and we don't want people to have to be Docker experts in order to start using Devcontainers. We believe that devcontainer features make it easier for people to add and maintain functionality in their configuration and encourage the use of features as a starting point. But if you have a scenario that needs more flexibility/control, you can certainly add a Dockerfile.

No, now you want people to be "feature" expert, and remember names, find exactly the thing they want. 
A simple "switch" to go back to the previous behavior should be released now.
I'll give you an example:

Just yesterday a colleague asked me a favor to help him fix a piece of python code, I use devcontainers exclusively and I went ahead and created a python one. And whilst installing the dependencies I noticed I had to add `libodbc.so.2`. Then I spend 30 minutes trying to figure out what feature has that and changes else so no other dependency breaks. I didn't manage, and had to manually create the DevContainer file.

Why was it so hard to keep the previous option whilst adding the new features on the side? Why the push towards "programming a json file", isnt better for everyone to understand how your dev environment work if they want?

I know Dockerfiles are still available, but the wizard to create them was REALLY nice, and now I have to manually map things.

But hey,maybe the top post's reactions might give a clue.

Cheers

**Comment by 7d4b9** [comment]
Dockerfile are by nature portable.
Feature are not and they have to make a lot filtering to be.
I prefer be a docker expert that a `if(platformA) else if (platformB)` or `if(Debian) else if (alpine) else if (Darwin)` etc syntaxes expert.
Docker solves a big part of the challenge by allowing to provide fine tuned images.
This is not the case with features that will become a pain to extend and support.
If someone is not a docker expert this is not a problem because the images are provided to him by docker expert usually.

**Comment by 7d4b9** [comment]
I work in a company with a lot security and we cannot use features because every thing is custom for us. 
As an example: it is way more simple to provide the "company approach docker" from an ops team that did already have a working image that extending the `docker-outside-docker` feature to target our internal boilerplates.

**Comment by 7d4b9** [comment]
Dockerfile are by nature portable.
Feature are not and they have to make a lot filtering to be.
I prefer be a docker expert that a `if(platformA) else if (platformB)` or `if(Debian) else if (alpine) else if (Darwin)` etc syntaxes expert.

If someone is not a docker expert this is not a problem because the images are provided to him by docker expert usually.
Docker solves a big part of the challenge by allowing to provide fine tuned images with everything inside
This is not the case with features that will become a pain to extend and support.

I work in a company with a lot security and we cannot use features because every thing is custom for us.
As an example: it is way more simple to provide the "company approach docker" from an ops team that did already have a working image than extending the docker-outside-docker feature to target our internal boilerplates.

**Comment by bherbruck** [comment]
@andreujuanc makes a lot of good points. When working in any higher-level language with C-bindings it is common to require external packages. Relying on a list of features can be handy if we need to run shell scripts during docker build, but for just installing packages...

I too have spend considerable time trying to install a list of packages in my own features, doing it in a json string in the devcontainer.json file is not easily maintainable.

Many already know docker. Adding another statement to `RUN apt-get install my-packages` is trivial and second-nature. Developing and debugging our own feature is time consuming (especially when permissions come into play).

**Comment by 7d4b9** [comment]
Features are not portable and they are faraway less powerful than docker. My choice is done.
Feature are just a script that can execute on every machine. I will never develop a such tool.
If the tool is there and work maybe I will use it but I will manage to prepare a fully working development image with all my team tools.
More also, feature like docker-in-docker are taking a while to install at devcontainer creation. Providing those software in a docker image costs nothing.
Cheers

**Comment by 7d4b9** [comment]
Feature are like a gas factory

**Comment by 7d4b9** [comment]
More again, in a lot of company like the one I work for are using internal Debian/docker repository and have to manage network specific stuff that make feature almost totally unusable.
Devcontainer is great but now they try to PIMP it so much that maybe the dream is to have developers that cannot handle their own system. Not my choice

**Comment by andreujuanc** [comment]
I also forgot to mention that most of the time I reuse the same dockerfile for my devcontainers, my cicd, and my k8s. So that way I know that my app will run 100% of the time.

**Comment by andreujuanc** [comment]
> If the goal is simple installation of packages, I think it easier to use Features for installing any package instead of writing a Dockerfile.
You think is easier, but consider someone who needs to search and find the correct feature for what they need.
I think the "linux" community have more content than the "feature" community, hence, easier to find solutions.

> For example, here is a Feature to install any package with `apt-get -y install --no-install-recommends`, which has been downloaded quite a few times. https://github.com/rocker-org/devcontainer-features/tree/main/src/apt-packages
Might be a hot take, but if you are developing inside a linux container, you should be familiar with the apt-get command, at least.
I understand we are trying to abstract things out, but then we must not forget about the "advanced" use cases, which is what we are arguing here. We are not asking for the "features" to go away, but to at least have an option to rollback to the previous behaviour. 


> Compare this to installing on a Dockerfile, which has the advantage of not having to worry about options like `--no-install-recommends` and deleting caches (e.g. by `rm -rf /var/lib/apt/lists/*`). Also, if we install on a Dockerfile, we have to modify all Dockerfiles if we need to modify a command, whereas a Feature is distributed from a single location, so we only have to modify one location.
If you have 100 projects using that feature, it will have to be re-run in all containers, on the other side, if you have a dockerfile you just need to download it. Done.



The problem here is that when a developer has a problem, it will have to search how to fix it with a feature, or, instead run a custom command, will will increase in size anyway with time. So just keeping the scripts within the dockerfile makes the most sense. Publishing docker images is also a thing, so people can reuse them within the org.

**Comment by bamurtaugh** [comment]
Hi everyone! Thanks so much for the discussion and honest feedback so far, as this is incredibly helpful to shaping the dev container spec and its resources.

I've opened an issue to discuss this topic and potential solutions: https://github.com/devcontainers/templates/issues/135. Please feel free to share your thoughts there, we'd really appreciate any and all feedback. Thank you!

**Comment by Studio766** [comment]
> It seems after `v0.16x.x` `Dockerfile`s are no longer included in the `.devcontainer` directory when using vscode's "Add devcontainer configuration files". These are critical for installing extra packages and customizing the development environment.
> 
> Is there some new functionality that deprecates this way of working with devcontainers?
> 
> ```
> .
> ├── .devcontainer
> │   └── devcontainer.json
> |       <-- Dockerfile used to be here
> ├── package.json
> └── src
>     └── index.js
> ```
